cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,473 @@
+
+![SPIFFE and SPIRE Cilium Agent using Delegated Identity API](./images/SPIFFE-SPIRE-cilium-agent-delegated.png)
+
+Then, when required (whether we do this at Endpoint creation or later is still TBD):
+1. All cilium-agents running a workload mentioned in an auth-enabled policy connect to their local spire-agent, and watch a set of SPIFFE labels that includes the labels for the relevant SVID. The cilium-agent identity is allowed to do Delegated Identity requests to the Workload API (WL API), so this is allowed.
+1. The spire-agent watches SVIDs on the SPIRE server, sees any updates to all relevant SVIDs, and passes them back to the cilium-agent.
+1. The cilium-agent then figures out where the request is coming from, and connects to the cilium-agent on the source node to perform the mTLS handshake. Because this  a mutual TLS, if it succeeds, then the workloads are authenticated.
+The destination node will also record that the workload is authenticated for reverse traffic.
+1. The mTLS handlers in each cilium-agent then pass the auth success to their local dataplane, by telling it that identity A is authenticated to identity B for some period of time (the lifetime of the certificates in the mTLS handshake).
+The mTLS handlers will also need to indicate the remote node, along with the auth type.
+For that lifetime, traffic between Identity A and Identity B will flow without further checks.
+
+#### SPIFFE installation steps and flow
+
+1. User creates a cluster, without CNI, per usual Cilium install process. 
+1. User installs Cilium, with mTLS enabled, SPIFFE auth will not work until the following conditions are met:
+    * SPIRE Server must be installed somewhere and configured correctly. (Bring-your-own-SPIRE-Server is an anticipated future need, but the initial implementation will use an in-cluster SPIRE server.)
+    * Per-node SPIRE Agent must be installed in the cluster and configured correctly. The SPIRE agent talks to the SPIRE server over the network, but all other communication with the Cilium Agent is via domain sockets shared on the host's filesystem.
+    * Cilium Agent SPIFFE Identity must be created as per SPIRE Agent config.
+    * Cilium Operator SPIFFE Identity must be created as per SPIRE Server config.
+    * (This document will be updated after [#23806](https://github.com/cilium/cilium/issues/23806) is done.)
+1. Cilium agent starts up as normal, acts as CNI.
+1. Cilium agent also contacts the local SPIRE agent at startup (via a domain socket shared on the host's filesystem) to watch the Delegated Identity API and gets its own SPIFFE identity via the SPIRE workload API.
+1. When generating Cilium Security Identities for identities with mTLS auth enabled, Cilium Operator in SPIFFE mode also records the SPIFFE identity (that is, the string `spiffe://spiffe.cilium.io/cilium-id/1337`).
```

Comment:
nit: double cilium is redundant in the suggested form. The _security identities_ here are typically referred to in the Cilium community either buy the full name, "security identity", or by the shorthand "identity". Occasionally in the datapath it may be "sec-id". The danger here is that "id" can end up colliding between multiple concepts, for instance Cilium Endpoints also have a node-local identifier known as the "endpoint id".

As such I'd suggest perhaps:

```suggestion
1. When generating Cilium Security Identities for identities with mTLS auth enabled, Cilium Operator in SPIFFE mode also records the SPIFFE identity (that is, the string `spiffe://spiffe.cilium.io/identity/1337`).
```

or:

```suggestion
1. When generating Cilium Security Identities for identities with mTLS auth enabled, Cilium Operator in SPIFFE mode also records the SPIFFE identity (that is, the string `spiffe://spiffe.cilium.io/security-identity/1337`).
```

(or a plural form of one of the above if we think that seems more approriate)

Reply:
I like the bare "identity" personally, I'll update to that.